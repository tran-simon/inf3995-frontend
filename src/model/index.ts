export type ObjectWithKey<T> = T & {
  _key: string;
};
