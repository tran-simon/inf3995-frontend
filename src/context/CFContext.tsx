import React, {
  createContext,
  ReactNode,
  useCallback,
  useEffect,
  useState,
} from 'react';
import Crazyflie from '../model/Crazyflie';
import { noop } from 'lodash';
import { SetState } from '../utils/utils';

const BACKEND_URL =
  process.env.REACT_APP_BACKEND_URL ?? 'http://localhost:5000';

interface ICFContext {
  cfList: Crazyflie[];

  scan: () => Promise<Response | void>;
  updateStats: () => Promise<Response | void>;

  backendUrl?: string;
  setBackendUrl: SetState<string | undefined>;

  backendDisconnected?: boolean;

  refreshRate?: number;
  setRefreshRate: SetState<number>;

  land: () => Promise<Response | void>;
  takeoff: () => Promise<Response | void>;
}

const DefaultCfContext: ICFContext = {
  cfList: [],
  scan: async () => {},
  updateStats: async () => {},
  backendUrl: BACKEND_URL,
  setBackendUrl: noop,

  setRefreshRate: noop,

  land: async () => {},
  takeoff: async () => {},
};

const CFContext = createContext<ICFContext>(DefaultCfContext);

export const CFProvider = ({
  children,
  ...props
}: Partial<ICFContext> & {
  children: ReactNode;
}) => {
  const [refreshRate, setRefreshRate] = useState<number>(
    props.refreshRate ?? 0,
  );
  const [backendUrl, setBackendUrl] = useState<string | undefined>(
    props.backendUrl ?? BACKEND_URL,
  );
  const [backendDisconnected, setBackendDisconnected] = useState(
    props.backendDisconnected ?? !!backendUrl,
  );
  const [cfList, setCfList] = useState<Crazyflie[]>(props.cfList || []);

  const scan = useCallback(async () => {
    if (!backendDisconnected) {
      return fetch(`${backendUrl}/scan`).then((res) => {
        if (res.ok) {
          res.json().then((value) => {
            if (Array.isArray(value)) {
              setCfList(value);
            }
          });
        }
      });
    }
  }, [backendUrl, backendDisconnected]);

  const updateStats = useCallback(async () => {
    if (!backendDisconnected) {
      return fetch(`${backendUrl}/updateStats`).then((res) => {
        if (res.ok) {
          return res.json().then((val) => {
            if (Array.isArray(val)) {
              setCfList(val);
            }
          });
        }
      });
    }
  }, [backendUrl, backendDisconnected]);

  useEffect(() => {
    fetch(`${backendUrl}/liveCheck`)
      .then((response) => {
        setBackendDisconnected(!response.ok);
      })
      .then(updateStats)
      .catch(() => {
        setBackendDisconnected(true);
      });
  }, [backendUrl, updateStats]);

  useEffect(() => {
    if (!backendDisconnected && !!refreshRate) {
      const refresh = setInterval(() => {
        updateStats();
      }, refreshRate);
      return () => clearInterval(refresh);
    }
  }, [refreshRate, backendUrl, backendDisconnected, updateStats]);

  const takeoff = async () => {
    if (!backendDisconnected) {
      return fetch(`${backendUrl}/takeOff`);
    }
  };
  const land = async () => {
    if (!backendDisconnected) {
      return fetch(`${backendUrl}/land`);
    }
  };

  return (
    <CFContext.Provider
      value={{
        ...DefaultCfContext,
        ...props,
        backendUrl,
        setBackendUrl,
        backendDisconnected,
        scan,
        updateStats,
        refreshRate,
        setRefreshRate,
        cfList,
        takeoff,
        land,
      }}
    >
      {children}
    </CFContext.Provider>
  );
};

export default CFContext;
