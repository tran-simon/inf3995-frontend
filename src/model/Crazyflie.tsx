import Point, { addPoint, newPoint, scalePoint } from '../utils/Point';
import Wall from './Wall';
import { SENSOR_MAX_RANGE } from '../utils/constants';

export default interface Crazyflie {
  battery?: number;
  speed?: number;
  state?: StateType;
  data: CFData[];
  initialPosition?: Point;
}

export interface CrazyflieDTO {
  droneId?: string;
  battery?: number;
  speed?: number;
  state?: StateType;
  cfData?: CFDataDTO;

  initialPosition?: Point;
}

type CFDataDTO = (string | undefined)[][];

export interface CFData extends Partial<Point>, Sensors {}

export const cfDataFromDTO = (cfDataDto: CFDataDTO): CFData => {
  const [x, y] = cfDataDto[0];
  const [north, south, east, west] = cfDataDto[1];

  return {
    x: x != null ? +x : undefined,
    y: y != null ? +y : undefined,
    north: north != null ? +north : undefined,
    south: south != null ? +south : undefined,
    east: east != null ? +east : undefined,
    west: west != null ? +west : undefined,
  };
};

export interface Sensors {
  north?: number;
  south?: number;
  east?: number;
  west?: number;

  [key: string]: number | undefined;
}

export type StateType = 'Standby' | 'In mission' | 'Crashed';
export const State = {
  standby: 'Standby' as StateType,
  inMission: 'In mission' as StateType,
  crashed: 'Crashed' as StateType,
};

export const getWalls = (crazyflie: Crazyflie): Wall[] => {
  const res: Wall[] = [];
  const { initialPosition, data = [] } = crazyflie;

  initialPosition &&
    data.forEach((data) => {
      const { east, north, south, west, x, y } = data;
      for (let i = 0; i < 4; i++) {
        let v: number | undefined;

        let xScale = 1,
          yScale = 1;
        switch (i) {
          case 0:
            xScale = 0;
            v = north;
            break;
          case 1:
            yScale = -1;
            xScale = 0;
            v = south;
            break;
          case 2:
            yScale = 0;
            v = east;
            break;
          case 3:
            xScale = -1;
            yScale = 0;
            v = west;
            break;
          default:
            break;
        }

        const notDetected = v == null || v >= SENSOR_MAX_RANGE;
        if (notDetected) {
          v = SENSOR_MAX_RANGE;
        }

        if (x != null && y != null) {
          res.push({
            cfData: {
              ...data,
              ...addPoint({ x, y }, initialPosition),
            },
            position: addPoint(
              addPoint({ x, y }, initialPosition),
              scalePoint(newPoint(xScale, yScale), v ?? SENSOR_MAX_RANGE),
            ),
            outOfRange: notDetected,
          });
        }
      }
    });

  return res;
};
