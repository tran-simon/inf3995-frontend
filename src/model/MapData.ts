import Crazyflie from './Crazyflie';
import { format } from 'date-fns';
import _ from 'lodash';
import { KeyArray } from '../utils';

export default class MapData {
  date: number;
  cfList: KeyArray<Crazyflie>;
  name?: string;
  simulation?: boolean;

  private constructor(mapData: MapData) {
    this.date = mapData.date;
    this.cfList = mapData.cfList;

    Object.assign(this, mapData);
  }

  /**
   * Create a new MapData with default values
   * date will be set to today
   * name will be the date formatted
   *
   */
  static create(mapData?: Partial<MapData>): MapData {
    const date = mapData?.date ?? Date.now();
    return new MapData({
      date,
      name: format(Date.now(), 'Pp'),
      cfList: {},
      ...mapData,
    });
  }

  /**
   * Creates a MapData from a DTO
   *
   * default values will be used:
   * date: The current date
   *
   * @param mapData
   */
  static fromDto(mapData: MapDataDTO): MapData {
    /* Add the required values (data) to all the cfs in the list */
    const cfList = !mapData.cfList
      ? {}
      : Object.keys(mapData.cfList).reduce(
          (cfList, key): KeyArray<Crazyflie> => {
            const cf = mapData.cfList && mapData.cfList[key];
            return !cf
              ? cfList
              : {
                  ...cfList,
                  [key]: {
                    ...cf,
                    data: cf.data ?? [],
                  },
                };
          },
          {},
        );

    return new MapData({
      date: Date.now(),
      ...mapData,
      cfList,
    });
  }

  /**
   * Remove undefined values from mapData to transform it into a Data Transfer Object ready to be saved in the database
   * @param mapData
   */
  static toDto(mapData: MapData): MapDataDTO {
    /* Removed undefined values from objects in cfList */
    const cfList = Object.keys(mapData.cfList).reduce(
      (cfList, key): KeyArray<Crazyflie> => {
        const cf = mapData.cfList[key];
        return !cf
          ? cfList
          : {
              ...cfList,
              [key]: _.omitBy(
                {
                  ...cf,
                  data: cf.data.map((data) => _.omitBy(data, _.isUndefined)),
                },
                _.isUndefined,
              ),
            };
      },
      {},
    );

    return _.omitBy<MapData>(
      {
        ...mapData,
        cfList,
      },
      _.isUndefined,
    );
  }
}

export type MapDataDTO = Partial<MapData>;
