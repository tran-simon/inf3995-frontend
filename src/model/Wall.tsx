import { Point } from '../utils/utils';
import Crazyflie from './Crazyflie';

export default interface Wall {
  position: Point;
  crazyflie: Crazyflie;
}
