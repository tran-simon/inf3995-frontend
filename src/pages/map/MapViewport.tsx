import React, { ReactNode, SVGProps, useContext } from 'react';

import CFContext from '../../context/CFContext';
import { getWalls, State } from '../../model/Crazyflie';
import Wall from '../../model/Wall';

const MapViewport = React.forwardRef<
  SVGSVGElement,
  Partial<SVGProps<SVGSVGElement>>
>((props, ref) => {
  const { cfList } = useContext(CFContext);

  const cfsSvg: ReactNode[] = [];
  const linesSvg: ReactNode[] = [];
  const wallsSvg: ReactNode[] = [];

  Object.keys(cfList).map((droneId) => {
    const addWalls = (walls: Wall[]) => {
      walls.forEach((wall, i) => {
        const origin = wall.cfData;
        const destination = wall.position;

        if (origin) {
          linesSvg.push(
            <line
              key={`line_${i}_${droneId}`}
              x1={origin.x}
              x2={destination.x}
              y1={origin.y}
              y2={destination.y}
              strokeWidth={3}
              stroke={'white'}
              strokeOpacity="0.8"
              strokeLinecap="round"
            />,
          );
        }
        if (!wall.outOfRange) {
          wallsSvg.push(
            <rect
              key={`wall_${i}_${droneId}`}
              width="1"
              height="1"
              {...destination}
              fill="red"
            >
              <title>Mur trouv√© par le drone: {droneId}</title>
            </rect>,
          );
        }
      });
    };

    const cf = cfList[droneId];
    if (!cf.data.length) {
      return;
    }

    const { x, y } = cf.data[cf.data.length - 1];
    if (cf.data) {
      cfsSvg.push(
        <circle
          key={droneId}
          r="1"
          cx={x}
          cy={y}
          fill="blue"
          stroke={cf.state === State.crashed ? 'red' : undefined}
        >
          <title>{droneId}</title>
        </circle>,
      );
    }
    addWalls(getWalls(cf));
  });

  return (
    <svg
      width="100%"
      height="100%"
      id="map-viewport"
      {...props}
      style={{ backgroundColor: 'grey', ...props.style }}
      ref={ref}
    >
      <defs>
        <filter id="f1" x="0" y="0" filterUnits="userSpaceOnUse">
          <feGaussianBlur in="SourceGraphic" stdDeviation="5" />
        </filter>
      </defs>
      {linesSvg}
      {wallsSvg}

      {cfsSvg}
    </svg>
  );
});

export default MapViewport;
