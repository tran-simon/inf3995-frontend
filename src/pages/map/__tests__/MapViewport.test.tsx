import MapViewport from '../MapViewport';
import { create } from 'react-test-renderer';
import CFContext from '../../../context/CFContext';
import React from 'react';
import { State } from '../../../model/Crazyflie';

describe('MapViewport', () => {
  const cf0 = {
    droneId: 'cf_0',
    position: {
      x: 10,
      y: 10,
    },
    state: State.inMission,
  };
  const cf1 = {
    droneId: 'cf_1',
    position: {
      x: 20,
      y: 10,
    },
    state: State.crashed,
  };

  const cfContext: any = {
    cfList: [cf0, cf1],
    walls: [
      {
        position: {
          x: 30,
          y: 10,
        },
        crazyflie: cf0,
      },
      {
        position: {
          x: 35,
          y: 10,
        },
        crazyflie: cf0,
      },
      {
        position: {
          x: 10,
          y: 15,
        },
        crazyflie: cf1,
      },
      {
        position: {
          x: 10,
          y: 20,
        },
        crazyflie: cf1,
      },
    ],
  };

  it('can match snapshot', () => {
    expect(
      create(
        <CFContext.Provider value={cfContext}>
          <MapViewport />
        </CFContext.Provider>,
      ),
    ).toMatchSnapshot();
  });
});
