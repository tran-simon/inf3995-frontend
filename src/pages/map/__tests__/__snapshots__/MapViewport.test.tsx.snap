// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapViewport can match snapshot 1`] = `
<svg
  height="100%"
  id="map-viewport"
  style={
    Object {
      "backgroundColor": "grey",
    }
  }
  width="100%"
>
  <defs>
    <filter
      filterUnits="userSpaceOnUse"
      id="f1"
      x="0"
      y="0"
    >
      <feGaussianBlur
        in="SourceGraphic"
        stdDeviation="5"
      />
    </filter>
  </defs>
  <line
    stroke="white"
    strokeLinecap="round"
    strokeOpacity="0.8"
    strokeWidth={3}
    x1={10}
    x2={30}
    y1={10}
    y2={10}
  />
  <line
    stroke="white"
    strokeLinecap="round"
    strokeOpacity="0.8"
    strokeWidth={3}
    x1={10}
    x2={35}
    y1={10}
    y2={10}
  />
  <line
    stroke="white"
    strokeLinecap="round"
    strokeOpacity="0.8"
    strokeWidth={3}
    x1={20}
    x2={10}
    y1={10}
    y2={15}
  />
  <line
    stroke="white"
    strokeLinecap="round"
    strokeOpacity="0.8"
    strokeWidth={3}
    x1={20}
    x2={10}
    y1={10}
    y2={20}
  />
  <rect
    fill="red"
    height="1"
    width="1"
    x={30}
    y={10}
  >
    <title>
      Mur trouvé par le drone: 
      cf_0
    </title>
  </rect>
  <rect
    fill="red"
    height="1"
    width="1"
    x={35}
    y={10}
  >
    <title>
      Mur trouvé par le drone: 
      cf_0
    </title>
  </rect>
  <rect
    fill="red"
    height="1"
    width="1"
    x={10}
    y={15}
  >
    <title>
      Mur trouvé par le drone: 
      cf_1
    </title>
  </rect>
  <rect
    fill="red"
    height="1"
    width="1"
    x={10}
    y={20}
  >
    <title>
      Mur trouvé par le drone: 
      cf_1
    </title>
  </rect>
  <circle
    cx={10}
    cy={10}
    fill="blue"
    r="1"
  >
    <title>
      cf_0
    </title>
  </circle>
  <circle
    cx={20}
    cy={10}
    fill="blue"
    r="1"
    stroke="red"
  >
    <title>
      cf_1
    </title>
  </circle>
</svg>
`;
